File: sources/perl/14_file_read/main.pl
# task 14 file_read — expected output: 484442112
# build: perl main.pl    run: perl main.pl
# note: data.bin is a fixture of 104857600 bytes, the bytes 0..255 repeating, opened in the working
#       directory and read in 1 MiB chunks, never one byte per syscall.
use strict;
use warnings;

my $chunk_size = 1048576;
my $total = 0;

open(my $fh, '<:raw', 'data.bin') or die "cannot open data.bin: $!";
while (my $got = read($fh, my $buf, $chunk_size)) {
    $total += $_ for unpack('C*', substr($buf, 0, $got));
}
close($fh);

print $total % 4294967296, "\n";

File: sources/python/14_file_read/main.py
# task 14 file_read — expected output: 484442112
# build: python main.py | pypy main.py | graalpy main.py | nuitka --standalone main.py    run: python main.py
# note: data.bin is a fixture of 104857600 bytes, the bytes 0..255 repeating; it is read from the working directory in 1 MiB chunks.

total = 0

with open('data.bin', 'rb') as f:
    while True:
        chunk = f.read(1 << 20)
        if not chunk:
            break
        for byte in chunk:
            total += byte

print(total % 4294967296)
